### ISSUE TITLE
Different performance between training and testing of AR task.

### ISSUE THREAD

**Issue by Zazexy** [fla-org/flash-linear-attention#553]
Thanks for your excellent work.  

I’ve encountered an issue while trying to use GLA in an autoregressive (AR) model. I’m pleased to see the speed improvements in both training and inference. However, when I tested GLA’s output with the same input in training and evaluation modes—where the input is `[b, n, h, d]` for training and `[b, 1, h, d]` (processed in a loop) for evaluation—the outputs turned out to be significantly different. The mean difference in outputs is around `1e-1`, and the maximum difference even reaches `1e1`. This stands in contrast to softmax attention, where the maximum difference is around `1e-1` and the mean is around `1e-2`.  

To troubleshoot this, I created a simple unit test using a transformer with stacked GLA layers, but the discrepancy remained unchanged. I suspect there might be bugs in my code. 

So, I’d like to ask if you have any testing code for verifying the differences between training and inference outputs.

**Comment by sustcsonglin** [comment]
Hi @Zazexy, this is a known issue with linear attention—there can be a noticeable discrepancy between the chunked (prefill) and recurrent (decoding) modes, especially under BF16 precision. If you're only concerned with final accuracy during evaluation, you might consider using the `fused_recurrent` kernel even for prefilling.

**Comment by sustcsonglin** [comment]
Related issue: https://github.com/fla-org/flash-linear-attention/issues/83,  https://github.com/state-spaces/mamba/issues/187

**Comment by sustcsonglin** [comment]
In general, the output decoding results are satisfactory, even when the output logit discrepancies are relatively large. It might also be interesting to try using an FP32 LM head during decoding, as recommended in Section 3.2 of the MiniMax-M1 report (https://arxiv.org/abs/2506.13585).

**Comment by Zazexy** [comment]
**Thanks for your patience!** However, I'm still confused about several issues since I'm a beginner to linear attention.

1. I'm sorry that I didn't provide enough information: my task is not a language task, which doesn't use prefill during inference; instead, it uses the full recurrent mode.
2. I read the section on MiniMax that you mentioned (https://arxiv.org/abs/2506.13585). Thanks again for your kindness, but **the outputs I compared are from the attention layers (or transformer layers) when I conduct unit tests**. I think this is not related to the decoder (or did you mean that use a FP32 decoder can reduce the difference?).
3. I'm grateful that you addressed the related issue, where delicate code was written to test the difference. So, **I followed their example** (https://github.com/state-spaces/mamba/issues/187#issuecomment-1964966374) **to test GLA with the following code to debug my code**; **however, I still get a somewhat "large" gap between the two outputs when I use BF16**.
4. Long story short, I want to know **how much difference is reasonable for logits in the range (-1, 1)when applying GLA as attention layers** , or maybe using a correlation metric like MiniMax to describe would be better? Because I found that the difference in outputs after 32 transformer layers(all of them are GLA) has a mean of nearly 0.12 (BF16).

**Thank you again for your patience and help throughout this process.** I would be truly grateful if you could share any thoughts or insights on these questions—it would mean a lot for my work.


```
from __future__ import annotations

import time
import torch
import matplotlib.pyplot as plt

from fla.layers.gla import GatedLinearAttention

from fla.models.utils import Cache

torch.manual_seed(42)
@torch.inference_mode()
def run():
    batch, length, hidden_size = 2, 6400, 16 * 128
    x = torch.rand(batch, length, hidden_size).to("cuda").to(torch.bfloat16)
    model = GatedLinearAttention(
            hidden_size=hidden_size,
            num_heads=16,
            use_output_gate=True,
            layer_idx=0,
            mode='chunk'
    ).to("cuda").to(torch.bfloat16)
    y1, _, _ = model(x)
    assert y1.shape == x.shape

    y2, _, _ = model(x)

    outs = []
    state_cache = Cache(0)
    for i in range(length):
        out, _, state_cache = model(x[:, i : i + 1, :], past_key_values=state_cache, use_cache=True)
        outs.append(out)
    y3 = torch.cat(outs, 1)

    print(torch.allclose(y1, y2))            # prints True
    print(torch.allclose(y2, y3))            # prints False
    print(torch.allclose(y1, y3))            # prints False
    print(torch.allclose(y1, y3, atol=1e-4)) # prints False

    print(f'diff y1 - y3: {torch.mean(torch.abs(y1 - y3))}') # diff y1 - y3: 0.000396728515625 with bf16
    print(f'diff y2 - y3: {torch.mean(torch.abs(y2 - y3))}') # diff y2 - y3: 0.000396728515625 with bf16
    print(f'diff y1 - y2: {torch.mean(torch.abs(y1 - y2))}') # diff y1 - y2: 0.0               with bf16

    # print(f'diff y1 - y3: {torch.mean(torch.abs(y1 - y3))}') # diff y1 - y3: 1.0078461798457283e-07 with fp32
    # print(f'diff y2 - y3: {torch.mean(torch.abs(y2 - y3))}') # diff y2 - y3: 1.0078461798457283e-07 with fp32
    # print(f'diff y1 - y2: {torch.mean(torch.abs(y1 - y2))}') # diff y1 - y2: 0.0                    with fp32

run()
```

**Comment by sustcsonglin** [comment]
I  think this output diff range should be normal

**Comment by Zazexy** [comment]
Thank you for your response!
